_posts/2017-01-13-graphite-scaling.md

```diff
@@ -0,0 +1,108 @@
+ 
+We did, in fact, do some I/O performance testing and tuning of the SSDs prior to placing the servers into production. However, we quickly found that the primary bottleneck was 100% CPU utilisation from the ``fio`` disk performance tool we used. The 3.7GHz CPUs were simply not fast enough to enqueue or dequeue I/O through the kernel and to the RAID array. Below is the best ``fio`` benchmark result we could obtain, with commentary:
+
+``fio``'s SSD test profile performs four tests in sequence – sequential read, random read, sequential write, and then random write. The block size was set to the default 4KB, and the fastest throughput we could obtain was with a queue depth of 128:
+ 
+```
+seq-read  : (g=0): rw=read, bs=4K-4K/4K-4K/4K-4K, ioengine=libaio, iodepth=128
+rand-read : (g=1): rw=randread, bs=4K-4K/4K-4K/4K-4K, ioengine=libaio, iodepth=128
+seq-write : (g=2): rw=write, bs=4K-4K/4K-4K/4K-4K, ioengine=libaio, iodepth=128
+rand-write: (g=3): rw=randwrite, bs=4K-4K/4K-4K/4K-4K, ioengine=libaio, iodepth=128
+```
+ 
+* Sequential read performance was 1876MB/sec. Read IOPS was 480K, and latency was 266 microsecs. User CPU of 21% was the primary bottleneck:
+ 
+```
+seq-read  : io=65536MB, bw=1876.4MB/s, iops=480337, runt= 34928msec
+lat (usec): min=64, max=13770, avg=266.26, stdev=99.45
+cpu       : usr=20.62%, sys=79.43%, ctx=424, majf=0, minf=1135
+```
+ 
+* Random read performance was 1771MB/sec. Read IOPS was 453K, and latency was 282 microsecs. User CPU of 24% was the primary bottleneck. Random read access for SSDs is not that much slower than sequential read performance:
+ 
+```
+rand-read : io=65536MB, bw=1771.2MB/s, iops=453401, runt= 37003msec
+lat (usec): min=64, max=13003, avg=281.99, stdev=56.98
+cpu       : usr=24.28%, sys=75.82%, ctx=262, majf=0, minf=4257
+```
+ 
+* Sequential write performance was 1718MB/sec. Write IOPS was 440K, and latency was 291 microsecs. User CPU of 21% was the primary bottleneck. Write performance looks to be very similar to read:
+ 
+```
+seq-write : io=65536MB, bw=1717.8MB/s, iops=439735, runt= 38153msec
+lat (usec): min=19, max=12445, avg=290.86, stdev=82.87
+cpu       : usr=20.98%, sys=79.01%, ctx=623, majf=0, minf=1004
+```
+ 
+* Random write performance was 1530MB/sec. Write IOPS was 392K, and latency was 326 microsecs. User CPU of 26% was the primary bottleneck. Random write performance was the “slowest”:
+ 
+```
+rand-write: io=65536MB, bw=1530.1MB/s, iops=391926, runt= 42807msec
+lat (usec): min=19, max=13972, avg=326.25, stdev=93.51
+cpu       : usr=25.68%, sys=74.31%, ctx=411, majf=0, minf=6857
+```
+# The Future
```

Comment:
This isn't rendering correctly for me, I think it just needs a newline before this heading